serverless/serverless
# Download templates from a Bitbucket Server

## What did you implement:

Allow the usage of templates hosted in a Bitbucket Server for creating a new Serverless service.
The authentication is managed using Basic Authentication as what was done for GitHub.



## How did you implement it:

In case a given URL do not refer to the already defined providers (GitHub, (public) Bitbucket, Gitlab), we try to fetch `/rest/api/1.0/application-properties` from the server. This endpoint in the Bitbucket Server Rest-API do not require any authentication. In case the response contains a property called `displayName` set to `Bitbucket`, we fulfill `repoInformation` with the data extracted from the URL for the Bitbucket Server. 


## How can we verify it:

Run unit-tests
Locally, check-out this branch. Create a repository with a Serverless template in a repository on your Bitbucket Server then run `serverless create -p test --template-url https://myname:mypass@myserver.com/rest/api/latest/projects/MY-PROJECT/repos/MY-REPO/archive?at=refs%2Fheads%2FsomeBranch


## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** NO  
**_Is it a breaking change?:_** NO
